cpu/cc2538/include/periph_cpu.h

```diff
@@ -45,12 +47,102 @@ typedef uint32_t gpio_t;
 /** @} */
 
 /**
+ * @name Internal GPIO shift and masking
+ * @{
+ */
+#define PORTNUM_MASK        (0x00007000)    /**< bit mask for GPIO port [0-3] */
+#define PORTNUM_SHIFT       (12U)           /**< bit shift for GPIO port      */
+#define PIN_MASK            (0x00000007)    /**< bit mask for GPIO pin [0-7]  */
+#define GPIO_MASK           (0xfffff000)    /**< bit mask for GPIO port addr  */
+/** @} */
+
+/**
  * @brief   Define a custom GPIO_PIN macro
  *
  * For the CC2538, we use OR the gpio ports base register address with the
  * actual pin number.
  */
-#define GPIO_PIN(port, pin) (gpio_t)(((uint32_t)GPIO_A + (port << 12)) | pin)
+#define GPIO_PIN(port, pin) (gpio_t)(((uint32_t)GPIO_A + \
+                                      (port << PORTNUM_SHIFT)) | pin)
+
+/**
+ * @brief Access GPIO low-level device
+ *
+ * @param[in] pin   gpio pin
+ *
+ * @return          pointer to gpio low level device address
+ */
+static inline cc2538_gpio_t *gpio(gpio_t pin)
+{
+    return (cc2538_gpio_t *)(pin & GPIO_MASK);
+}
+
+/**
+ * @brief   Helper function to get port number for gpio pin
+ *
+ * @param[in] pin   gpio pin
+ *
+ * @return          port number of gpio pin, [0=A - 3=D]
+ */
+static inline int gpio_port_num(gpio_t pin)
+{
+    return (int)((pin & PORTNUM_MASK) >> PORTNUM_SHIFT) - 1;
+}
```

Comment:
I can't reproduce that. I added debug output to `gpio_init`, you can try `tests/periph_gpio` for example and init one of the LEDs with `init_put 3 4` you will see debug output showing real pin address (uint32_t) as well as  port and pin recovered from that. For me looks like that:

```
> init_out 3 4
2017-07-06 16:52:33,730 - INFO #  init_out 3 4
2017-07-06 16:52:33,731 - INFO # GPIO 1074642948, PORT: 3, PIN: 4
```